cli/popper/utils.py

```diff
@@ -104,19 +116,112 @@ def is_popperized(rootfolder):
     return os.path.isfile(config_filename)
 
 
+def setup_logging(level='INFO'):
+    root = logging.getLogger()
+    f = colorlog.ColoredFormatter('%(log_color)s%(levelname)s: %(message)s ',
+                                  log_colors=log_colors)
+    ch = logging.StreamHandler(sys.stdout)
+    ch.setFormatter(f)
+    root.addHandler(ch)
+    set_level(level)
+
+
+def set_level(level):
+    levels = {
+        'DEBUG': logging.DEBUG,
+        'INFO': logging.INFO,
+        'WARNING': logging.WARNING,
+        'ERROR': logging.ERROR,
+        'CRITICAL': logging.CRITICAL
+    }
+    root = logging.getLogger()
+    root.setLevel(levels[level])
+
+
+def remove_level():
+    handler = logging.getLogger().handlers[0]
+    f = colorlog.ColoredFormatter('%(log_color)s%(message)s ',
+                                  log_colors=log_colors)
+    handler.setFormatter(f)
+
+
+def add_level():
+    handler = logging.getLogger().handlers[0]
+    f = colorlog.ColoredFormatter('%(log_color)s%(levelname)s: %(message)s ',
+                                  log_colors=log_colors)
+    handler.setFormatter(f)
+
+
+def multiline(log):
+    """
+    A decorator for multiline log.
+    This solves a few problems for multiline log.
+    1. Without the decorator, a multiline log will have coloring for the first line
+    2. Only the first line should have level information
+        Eg. log('This is a multiline\nline') should print
+         LEVEL:
+         This is a multiline
+         line log
+        Note: LEVEL is printed on a separate line, so that log can retain indent info
+              This is ugly.
+              LEVEL:    Each line in
+                  this log is indented
+                  by four
+                  spaces
+    Note: 'INFO' level doesn't have color and level. So decorator is not required
+    """
+    def wrapper(msg):
+        # Check if it is a multiline log
+        msg_lines = msg.splitlines(True)
+        pos = 0
+        line_count = 0
+        for line in msg_lines:
+            if not line.isspace():
+                line_count += 1
+        if line_count > 1:
+            if msg_lines[0].isspace():
+                remove_level()
+                for index, line in enumerate(msg_lines):
+                    if line.isspace():
+                        log(line)
+                    else:
+                        pos = index
+                        break
+            if pos < len(msg_lines):
+                add_level()
+                log('')
+                remove_level()
+                rest = ''.join(msg_lines[index:])
+                log(rest+"\n")
+        else:
+            log(msg)
+    return wrapper
+
+
+@multiline
 def fail(msg):
     """Prints the error message on the terminal."""
-    click.secho('ERROR: ' + msg, fg='red', bold=True, err=True, nl=False)
+    log.error(msg[:-1])
```

Comment:
If we remove the `fail`, `warn`, `info` wrappers, we can assume that the code that will call `logging` functions will **not** be ending messages new lines

Reply:
we need wrappers for `fail` because we need to call `sys.exit(1)` 
and for `info` as well because i am wrapping it inside `remove_level()` and `add_level()`
Without utilizing those functions, if we retain `levelname` for `info` it will be like [this](https://travis-ci.org/systemslab/popper/jobs/518905794#L658)

Reply:
Sorry @akashrajr1 , I briefly mentioned it in one of the earlier comments but I should have explained better what I meant.

The logging module has facilities for implementing custom handlers that change the default behavior. This helps us so that we don't need to wrap log calls, rather we provide a custom handler that the logging module makes use of.

With this, we can for example set the behavior in such a way that for INFO level messages, no prefix is shown; and that for FAIL, a message is shown and then an exception is raised (or sys.exit(1) is invoked).

Reply:
Thanks @ivotron, so the way forward is we implement our own custom handlers for `sys.exit(1)` and no `levelname` for `info`?
I was also trying a different approach by created a `log.py` module so that we can use `log.info()` directly. I still don't understand what might be the cons of using wrapping log calls.

Reply:
yeah, the handler facility from `logging` is the way a python codebase defines custom logging behavior. Wrapping log calls has the potential of introducing performance overheads, e.g. calling the functions that change levels every time we log a message might not be negligible.

Reply:
Thanks that clear things up